### PR TITLE
postmerge: re-derive the skill-vs-code claims #386 invalidated, and land #368 under its token bar

### PR DESCRIPTION
**Pull request by OsherElhadad**
Post-merge review of four skill PRs merged without independent review — **#368** agent-optimize, **#369** gepa, **#380** skillopt, **#373** mcp-tool — plus the fixes for everything real that it found. Full write-up in `.review/postmerge-algos.md`.

The headline: **#369 and #380 were honest when written and have decayed since.** Both re-derived their SKILL.md against the code, then `#386` (the shared `OptimizerContext`) and `#350` restructured `gepa.py`/`skillopt.py` underneath them. Nothing noticed, because a hand-derived skill-vs-code table has no guard. That is the failure mode this whole effort was fixing, reappearing not through carelessness but through absence of a mechanism.

## Re-derived divergence counts

| | claimed | re-derived on current `main` |
|---|---|---|
| **#369 gepa** | 14 divergences, 4 flagged as gaps | **2 of the 4 flagged gaps are no longer real**; 3 live divergences remain (1 of them a wrong budget ceiling), + 1 in `references/concepts.md` |
| **#380 skillopt** | 10 divergences, 5 flagged as gaps | **all 5 gaps still real** (one worse than documented), but **11 of 18 `file:line` citations now point at unrelated code** |

### gepa

Removed two gaps that are fixed:

- Candidate snapshots **do** pass `_SNAPSHOT_IGNORE` at both call sites (`gepa.py:682`, `:846`) since #350, and `NON_CAPABILITY_FILES`/`DIRS` **do** exclude `CLAUDE.md`/`AGENTS.md`/`.claude`/`guidance`/`trajectories`/`prior_iterations` since #386 (`types.py:26-43`) — so round-robin *cannot* burn an iteration on a dotfile. `references/concepts.md:100-105` asserted the exact opposite of what #386 deliberately changed.
- Optimizer-context parity with hill-climb (#109) is fixed: `gepa_loop` takes the shared `OptimizerContext` and calls the same `inject()`/`instructions()` as its siblings. Rewritten down to the half that *is* still true — `JOURNAL.md` never accumulates in gepa, because `_reconcile_journal` is called from `harness.run_step` only (`harness.py:1578`) and gepa runs the optimizer itself.

Fixed a budget number that could cost real money: the `--max-metric-calls` overshoot bound said "`2·mb + |val|·n-trials`, one more minibatch on a merge iteration". A merge fires *inside* an accepting iteration and evaluates the merge **and both parents** on a **freshly sampled** minibatch, then pays a second full val — worst case `5·mb + 2·|val|·n_trials`.

Everything else in gepa's body verified accurate: the frequency-weighted per-instance sampling vs the strict `pareto_frontier` (two genuinely different sets), the `sum(child) > sum(parent)` local gate, the `reward < 1.0` threshold, the ~800-char/12-task caps, `--max-merges` counting attempts-that-built (a skip is free), the INDECISIVE tamper path, and all nine hyperparameter defaults. Gap 1 (no task input in the reflective dataset, despite the docstring at `gepa.py:222` claiming one) and gap 5 (empty train ⇒ silent **val** fallback at `:522`) are both still live.

### skillopt

All five gaps reproduce. But every citation that makes them checkable had drifted 7-9 lines: `skillopt.py:230`, cited for the train-ids read, is now a docstring line. All 11 refreshed, plus a standing note that a citation which stops matching means the section needs re-deriving. Also dropped the stale "*both land with PR #370; until then read `harness.run_step` itself*" — #370 landed as `ae71aa87`.

One gap was **understated**. `applied_changes` is not merely uninformative; its ignore list matches seven `.md` *basenames* while `run_step` injects `guidance/`, `trajectories/`, `prior_iterations/*/diff.patch` that `_SNAPSHOT_IGNORE` keeps *out* of the parent snapshot. Measured zero-API on `examples/toy_calc` at `L=4`:

```
epoch 1 step 1  requested_edits=4  applied_changes=10  accept=true
epoch 2 step 1  requested_edits=4  applied_changes=10  accept=false
```

Step 1's 10 = 7 injected `guidance/*` + 2 `trajectories/*` + **1** real edit. Step 2 logged 10 again with `prompt.txt` **byte-identical to its parent**. The metric has no zero, so it cannot detect the one thing it exists to detect: an optimizer that made no edit.

One imprecision corrected: gap 1 said the step renders "no task ids". The focus summary and failure index are empty, but `_passing_block` is built from the **whole** val split since #370 (`harness.py:2013`), so protect-these-ids *is* populated — and it is the only per-task content a step gets.

## #368's token budget

Measured with `tiktoken` `cl100k_base`, body only (the `description` is separately-budgeted always-in-context metadata, per `.review/skill-duplication-analysis.md`):

| | lines | body tokens |
|---|--:|--:|
| pre-#368 | 922 | — |
| as merged by #368 | 337 | **5,580** |
| this PR | 311 | **5,000** |

#368 reported ~5,444; the real figure was **5,580**, 11.6% over the ≤5,000 bar, flagged rather than fixed. **It was not earning its place, so it is cut** — −580 tokens, no rule deleted. What moved was restatement of its own references or of `references/algorithm.md`: the fan-out invariants' rationale (`algorithm.md` § *Parallelism* states all four in full), the parallelism enumeration, the helper flags `per-task-fanout.md` carries verbatim (`--supersedes`, `--floor`, `--canary-auto`, `--traces`, `dropped_additions`), the `measure.py` refusals' reasoning. Plus one genuine triplication: "a candidate inside the null band is not evidence" appeared three times in one body.

Two counting bugs #368 left behind, both fixed: "state these **five** invariants" over a four-item list (it had merged items 4+5), and "**Two** rules decide whether a round's number means anything" over three.

**One real instruction loss, restored.** SKILL.md names "the sign test for effects below the floor" as living in `references/measured-lessons.md`. #368 moved the section and dropped the rule, so the pointer led nowhere. Pre-registered directional predictions + a sign test (9/10 positive, `p = 0.0107`, where no individual z reached 2) are back where the pointer promises them, with the two conditions that make it a test rather than a story.

`core/tests/test_agent_optimize_skill.py` **passes unmodified** — and it earned its keep mid-review: it pins the literal term `re-gate`, which one of my compressions had paraphrased away. `scripts/check.py` caught a second (`no-regression`). Both restored. Worth naming rather than quietly fixing.

## #373's guard, verified empirically

Reproduced against the shipped `tool_surface` with mcp-tool's real policy:

```
(1) params edit carrying properties/type/required
    report: {"changed": ["params:kb_search"], "refused": []}
    parameters now: {"type": "array", "properties": {...}, "required": ["injected"]}
(2) add edit carrying a code key
    report: {"changed": ["add:srv"], "refused": []}   added keys: ['code','description','name','parameters']
(3) validate() → {"ok": true, "problems": []}         references load_policy: False
(4) only inputs/policy.json  → ignored
    capability_dir/policy.json → read
```

All four claims hold, and the skill states them plainly under "**The policy checks the edit's LABEL, not its effect**" rather than implying enforcement. **No change needed to the skill.**

But #373's policy-path correction had not reached the code that *defines* the answer: `tool_surface.py`'s module docstring and `load_policy`'s own docstring both still told the reader the loader reads `inputs/policy.json`, which it ignores. Fixed here (core, not a skill).

**No skill in scope names the wrong path.** `mcp-tool/SKILL.md:58-61` mentions `inputs/policy.json` only to rule it out.

## Instruction loss

`git show` on each merge commit, removed lines diffed against the current body + every reference, then a mechanical audit of bolded spans and backticked identifiers. Beyond #368's sign test: **#369 no loss**, **#373 no loss** (the annotations table, human-in-the-loop, and execution-vs-protocol errors all landed in `references/concepts.md` §4). Ownership compliance is good across all four — nobody restates the honesty invariants, the agent-mode loop, the failure taxonomy or the reflective-dataset format. Two pointer shortfalls fixed: gepa pointed at `hill-climb/SKILL.md` rather than its `references/run-step.md`, and skillopt's pointer was gated on an already-landed PR.

## The test convention this adds

`core/tests/test_skill_code_claims.py` — 7 tests, 3 **failing-first** (proof: `/tmp/capreview-postmerge-algos-failingfirst.py`). The point is the convention, not the coverage:

> **A skill that documents a code defect needs a test that fails when the defect is FIXED.**

Otherwise the doc silently becomes a lie the moment someone does the right thing — which is exactly what #386 did to #369, twice. `test_skillopt_minibatch_focus_is_still_empty_by_construction` fails the day #371 lands, and its message names the paragraph to delete.

## Verification

- `PYTHONPATH="$PWD/core:$PWD/dashboard/backend" python -m pytest core/tests -q` → **812 passed, 12 skipped** (805 baseline + 7 new). Always pinned — #349.
- All four skills' `scripts/check.py` → `ok: true, problems: []`.
- Zero-API runs on `examples/toy_calc` with the mock optimizer, both reaching a gate-accepted candidate and a sealed test score:

  | | gepa | skillopt |
  |---|---|---|
  | best_id | `gepa_0001` | `so_e01s01` |
  | accepts / val | 1 / 1.0 | 1 / 1.0 |
  | **sealed test** | **1.0** (baseline 0.0) | **1.0** (baseline 0.0) |

  **Honest caveat:** both fired `gate_warning` — "paired SE is 0 (likely n_trials=1) … falling back to STRICT (accept any delta>0)". Expected on a deterministic `n_trials=1` example and reported honestly by the engine, but these are STRICT-fallback accepts. They prove the plumbing and the seal, not the gate's discrimination.

## What other owners should pick up

Not touched, per the ownership contract:

- **`tools`** — `references/{examples.md:366, pitfalls.md:127, concepts.md:114}` still say the effective policy is `inputs/policy.json`. Child of #352; the core docstrings are fixed here, so these three are the last wrong surfaces.
- **`phases/gate`** — #380 deleted a *"Pitfall: small val sets"* whose `--n-trials` half survives in skillopt's body but whose other half does not live anywhere: *use a **graded** reward so the paired significance test has signal.* `gate` owns the significance bar; it belongs there.

### CODE REVIEW

**Comment by OsherElhadad** [comment]
## Rebased onto `c8bc2292` — re-derived, not reused

`#387`, `#388`, `#391`, `#393`, `#396` all landed after I branched. Re-derived from scratch rather than reusing my earlier numbers, per the instruction — and that was the right call, because **three of my own conclusions changed and one of my own tests caught a mistake I had made.** Recorded here rather than quietly corrected, since the finding of this whole review is that a hand-derived skill-vs-code table decays.

### gepa's Known gaps: 5 bullets → 2

- **`#387` fixed the hollow eval-cache reflection** (`cache.py:91-99` now persists `output`/`trace`/`errored`; `gepa.py:162-175` replays them on a hit).
- **`#396` fixed BOTH the missing `step` record and the non-accumulating `JOURNAL.md`** — `harness.record_iteration` (`harness.py:1037-1077`) is now the single place any algorithm ends an iteration, and gepa calls it at four sites.

Two gaps survive, both re-verified: the reflective dataset still carries **no task input** while its own docstring says it does (`gepa.py:230` vs `:263-267`), and an empty `splits.train` still falls back to **val** ids silently (`:530`).

Also corrected a body claim in the opposite direction: I had left `JOURNAL.md` out of the injected cross-iteration files. It was omitted *because* it never accumulated — which `#396` fixed — so all five files are now real, and the skill says so.

### Two `harness.py` citations rotted again

`#393` + `#396` grew the file ~266 lines: `harness.py:2004-2006`/`:2013` → `:2270-2272`/`:2279`. Re-derived **mechanically** this time rather than by hand — **23/23 skillopt + 3/3 gepa citations resolve**. All 19 `skillopt.py` citations held; only the cross-file ones moved, which is the predictable failure mode and a decent argument for citing symbols over line numbers next time.

### My own tripwire fired — right reason, wrong cause

`test_skillopt_minibatch_focus_is_still_empty_by_construction` went red. Not because `#371` was fixed, but because **`#391` reworded** the summary from `of 0 tasks` to `of 0 focused task(s) of N on val`. The gap is exactly as real.

Two fixes, both mine to own: the test now asserts the **property** (`0 solid / 0 flaky / 0 failing`, plus no val feedback leaking into a train-focused prompt) instead of a sentence, and skillopt's SKILL.md no longer quotes the pre-`#391` string. A tripwire that fires on rewording is a false alarm, which is worse than none.

### Preserved exactly

- **The gepa budget-ceiling correction**, re-verified against current `gepa.py`: `_try_merge` samples a **fresh** minibatch (`:831`) and evaluates the merge plus **both** parents (`:832`/`:834`/`:836`) before a second full val. `5·mb + 2·|val|·n_trials` stands.
- **`#396`'s AST guard** in `core/tests/test_iteration_record_parity.py` — untouched (`git diff` empty, 9 passed).
- **`#388`'s fix.** Both merge conflicts were `#388` removing the phantom `cap-evolve status` / `cap-evolve finalize` subcommands, colliding with my token trims. `#388`'s substance won; the phantom commands were **not** reintroduced (verified against the 14 subcommand-guard tests), and I kept its Stop-hook sentence, finding the tokens elsewhere.

### Added from `#396`'s semantics

An INDECISIVE gepa child **charges `spent.iterations`** while leaving the stall counter alone (`record_iteration(..., indecisive=True)`, `harness.py:1072`). The spend meter counts it because the rollouts were really spent; only the evidence meter does not. A reader budgeting a run needs that distinction and the skill did not have it.

### Evidence

```
PYTHONPATH="$PWD/core:$PWD/dashboard/backend" python -m pytest core/tests -q
  → 836 passed, 12 skipped          (main's 829 + my 7)

scripts/check.py    agent-optimize ok []   gepa ok []   skillopt ok []   mcp-tool ok []
body tokens         agent-optimize 5000 (311 lines) · gepa 2431 · skillopt 2276 · mcp-tool 1840
citations           23/23 skillopt + 3/3 gepa resolve
failing-first       all 3 re-verified RED against origin/main

bash examples/toy_calc/run.sh
  → best_id cand_0001 · baseline_val 0.0 · best_val 1.0
    test_reward 1.0 · test_delta 1.0 · finalized true · iterations 3
```

`toy_calc` carries the same caveat as before: at `n_trials=1` on a deterministic example, paired SE collapses to 0 and `gate.py` applies its documented STRICT fallback. It proves the plumbing, the split and the seal — not the gate's discrimination.

Updated in place (`MERGEABLE`); no superseding PR needed.
